### PR TITLE
[SPARK-47655][SS] Integrate timer with Initial State handling for state-v2

### PR DESCRIPTION
**Pull request by jingz-db**
### What changes were proposed in this pull request?

We want to support timer feature during initial state handling. Currently, if grouping key introduced in initial state is never seen again in new input rows, those rows in initial state will never be emitted. We want to add a field in initial state handling function such that users can manipulate the timers with initial state rows.

### Why are the changes needed?

These changes are needed to support timer with initial state. The changes are part of the work around adding new stateful streaming operator for arbitrary state mgmt that provides a bunch of new features listed in the SPIP JIRA here - https://issues.apache.org/jira/browse/SPARK-45939

### Does this PR introduce _any_ user-facing change?

Yes. We add a new `timerValues` field in `StatefulProcessorWithInitialState`:
```
  /**
   * Function that will be invoked only in the first batch for users to process initial states.
   *
   * @param key - grouping key
   * @param initialState - A row in the initial state to be processed
   * @param timerValues  - instance of TimerValues that provides access to current processing/event
   *                     time if available
   */
  def handleInitialState(key: K, initialState: S, timerValues: TimerValues): Unit
```

### How was this patch tested?

Add unit tests in `TransformWithStateWithInitialStateSuite`

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala [hunk 1]
```diff
@@ -125,6 +129,29 @@ class InitialStateInMemoryTestClass
   }
 }
 
+class StatefulProcessorWithInitialStateProcTimerClass
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala:132]: nit: could we add a small comment for each of these classes ? i.e. to explain their behavior for the initial state methods ?

File: sql/api/src/main/scala/org/apache/spark/sql/streaming/StatefulProcessor.scala [hunk 2]
```diff
@@ -107,6 +107,8 @@ trait StatefulProcessorWithInitialState[K, I, O, S] extends StatefulProcessor[K,
    *
```
Comment (anishshri-db) [inline sql/api/src/main/scala/org/apache/spark/sql/streaming/StatefulProcessor.scala:107]: In the line above - could we please update to include package private for sql ?

`private[sql] trait StatefulProcessorWithInitialState` ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala (hunk 1 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala:136]: nit: maybe we should clarify - that we are trying to test for how initial state, timers and outputting rows might interact ? not sure if thats clear from the comment above ?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala [hunk 3]
```diff
@@ -125,6 +129,43 @@ class InitialStateInMemoryTestClass
   }
 }
 
+/**
+ * Class to test stateful processor with initial state and processing timers.
+ * Timers can be registered during initial state handling and can be emitted after timer
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala:134]: Nit: `output rows can be emitted`

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala [hunk 4]
```diff
@@ -125,6 +129,43 @@ class InitialStateInMemoryTestClass
...
+      timerValues: TimerValues): Unit = {
+    // keep a _countState to count the occurrence of grouping key
+    // Will register a timer if key "a" is met for the first time
+    processUnexpiredRows(key, 0L, 1L, timerValues)
+  }
+}
+
+/**
+ * Class to test stateful processor with initial state and event timers.
+ * Timers can be registered during initial state handling and can be emitted after timer
```
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala:152]: same here

**Review by anishshri-db — APPROVED** [review]
lgtm - pending minor nits

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**